identity-did/src/revocation/bitmap.rs

```diff
@@ -0,0 +1,158 @@
+// Copyright 2020-2022 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use core::fmt::Formatter;
+use core::fmt::Result as FmtResult;
+use std::io::Write;
+
+use flate2::write::ZlibDecoder;
+use flate2::write::ZlibEncoder;
+use flate2::Compression;
+use identity_core::utils::decode_b64;
+use identity_core::utils::encode_b64;
+use roaring::RoaringBitmap;
+use serde::de;
+use serde::de::Deserializer;
+use serde::de::Visitor;
+use serde::ser::Error as _;
+use serde::ser::Serializer;
+use serde::Deserialize;
+use serde::Serialize;
+
+use crate::error::Error;
+use crate::error::Result;
+
+/// A compressed bitmap for managing credential revocation.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct RevocationBitmap(RoaringBitmap);
+
+impl RevocationBitmap {
+  /// Creates a new [`RevocationBitmap`].
+  pub fn new() -> Self {
+    Self(RoaringBitmap::new())
+  }
+
+  /// Returns `true` if the credential at the given `index` is revoked.
+  pub fn is_revoked(&self, index: u32) -> bool {
+    self.0.contains(index)
+  }
+
+  /// Revokes the credential at the given `index`.
+  ///
+  /// Return whether the value was absent from the set.
+  pub fn revoke(&mut self, index: u32) -> bool {
+    self.0.insert(index)
+  }
+
+  /// The credential at the given `index` will be set to valid.
+  ///
+  /// Returns ture is the value was present in the set.
+  pub fn undo_revocation(&mut self, index: u32) -> bool {
+    self.0.remove(index)
+  }
+
+  /// Deserializes a compressed [`RevocationBitmap`] base64-encoded `data`.
+  pub fn deserialize_compressed_b64(data: &str) -> Result<Self> {
+    let decoded_data: Vec<u8> = decode_b64(data).map_err(|e| Error::Base64DecodingError(data.to_owned(), e))?;
+    let decompressed_data: Vec<u8> = Self::decompress_zlib(decoded_data)?;
+    Self::deserialize_slice(&decompressed_data)
+  }
+
+  /// Serializes and compressess [`RevocationBitmap`] as a base64-encoded `String`.
+  pub fn serialize_compressed_b64(&self) -> Result<String> {
+    let serialized_data: Vec<u8> = self.serialize_vec()?;
+    Self::compress_zlib(&serialized_data).map(|data| encode_b64(&data))
+  }
+
+  /// Deserializes [`RevocationBitmap`] from a slice of bytes.
+  pub fn deserialize_slice(data: &[u8]) -> Result<Self> {
+    RoaringBitmap::deserialize_from(data)
+      .map_err(Error::DeserializationError)
+      .map(Self)
+  }
+
+  /// Serializes a [`RevocationBitmap`] as a vector of bytes.
+  pub fn serialize_vec(&self) -> Result<Vec<u8>> {
+    let mut output: Vec<u8> = Vec::with_capacity(self.0.serialized_size());
+    self.0.serialize_into(&mut output).map_err(Error::SerializationError)?;
+    Ok(output)
+  }
```

Comment:
Not sure if we should be exporting many of these serialize/deserialize methods as public functions.